### PR TITLE
adding Multinest sampler

### PR DESCRIPTION
**Pull request by qyx268**
This is related to Issue #25 where we've added the multinest sampler

@steven-murray 
I don't see a decent way to output data (e.g. global_xH or PS) using multinest. Unlike cosmoHammer, I think it's not very straightforward to pass a storageUtil class. Do you have any suggestion? I'm thinking of outputting inside the likelihood function when calling multinest?

### CODE REVIEW

**Comment by qyx268** [comment]
image

Test with LikelihoodNeutralFraction, LikelihoodGreig, LikelihoodPlanck. Blue and Red correspond to Multinest and Emcee samplers, respectively. The posteriors look similar between the two. 

MultiNest has 582 total likelihood evaluations before reaching convergence with ln(ev)=  -2.09      +/-  0.12. The Emcee result shown in the posterior has 9000 evaluations. But from the chain, it looks like it has converged after 6000 evaluations.

File: src/py21cmmc/likelihood.py
Comment (steven-murray) [inline src/py21cmmc/likelihood.py]: To keep with Python style guidelines, these should be all lower-case

File: src/py21cmmc/likelihood.py
Comment (steven-murray) [inline src/py21cmmc/likelihood.py]: change this variable to ``freq_tb_min``.

File: src/py21cmmc/likelihood.py
Comment (steven-murray) [inline src/py21cmmc/likelihood.py]: again, change all variables to be lower case, except for class names

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [inline src/py21cmmc/mcmc.py]: The "Essential" multinest options should be put in here explicitly, and the rest can be caught by the ``**multinest_options`` here.

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [inline src/py21cmmc/mcmc.py]: The parameters here should be added to the ``Parameters`` section (as noted in previous comment)

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [inline src/py21cmmc/mcmc.py]: do this as 

```python
try:
    import pymultinest
except ImportError:
    raise ImportError("You need to install pymultinest to use this function!")
```

Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: Also, add an ``extras_require`` in ``setup.py`` something like:

```python
setup(
    ...
    extras_require = {
        "samplers": ['multinest'],
    }
)
```

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [inline src/py21cmmc/mcmc.py]: From here to line 299 is very similar to `run_mcmc`, and we should probably refactor so we don't have redundant code.

**Review by steven-murray — COMMENTED** [review]
This is a great addition, thanks @qyx268 !

Your question about writing out "blobs" is basically the reason I haven't tried to do this myself in the past -- it seems to me like a tricky question. In fact, I'm not even really happy with the way that `emcee` writes out blobs right now (I've had to copy in code from `emcee v3` but modify it to get it to work). It's method of course only works for `emcee`, since it happens in the sampler itself, so it's not translatable to other samplers. 

We could imagine using a similar HDF5-backed system, but including it explicitly in the likelihood, so that other samplers work with it, as you suggest. However, I think that will take a little bit of thought, and we will want to get it basically right first time (so that we don't have to change the API too much). 

We have two options: (1) we can think about how to get the blobs right for different samplers right here in this PR, which will delay the acceptance of this PR, but means the blob saving will hopefully get into 21CMMC v1; (2) we can accept this PR as-is, and treat the proper writing out of blobs as an extra feature, for which we should make an issue, and postpone it to 21CMMC v1.1 or even v2. 

I'll invite @BradGreig and @caw11 to comment on which option they think is best.

File: src/py21cmmc/mcmc.py
Comment (qyx268) [reply src/py21cmmc/mcmc.py]: I don't think I understand this. These parameters are taken directly from MultiNest, which should be specified when calling run_multinest. Therefore I consider them as default Parameters, same as core_modules, likelihood_modules, etc. Putting them in a second section (Essential MultiNest Parameters) is just showing that they are from MultiNest.

Are you suggesting removing the title (Essential MultiNest Parameters) and put these parameters in the "Parameter" section, or putting these parameters in **essential_multinest?

File: src/py21cmmc/likelihood.py
Comment (qyx268) [reply src/py21cmmc/likelihood.py]: These are for #19.

File: src/py21cmmc/mcmc.py
Comment (qyx268) [reply src/py21cmmc/mcmc.py]: Are you suggesting simply renaming them or putting run_multinest as an option for run_mcmc?

**Comment by qyx268** [comment]
I would leave the decision to you whether to include this PR in the following release or to postpone it when outputting functions are included. I'm okay with both. But as you can see, this branch (for PR #27 ) was created from the mini-halos branch (for PR #19 ) where EDGES likelihood was included. I would suggest accepting #19 first and include EDGES likelihood in the v1. 

I also feel that the best way would be outputting directly in the likelihood. This is what I had in mind but relevant development would take much longer. 

I'm pushing this implementation because we might have a 2-month lifetime computing resource in a few weeks where I'd like to test the scalability of this sampler. For this, I might simply output tiny files for blobs as a temporary solution.

**Comment by qyx268** [comment]
Here is an updated comparison with 1000 lives points when running MultiNest (compared to 100 last time). It has 5892 evaluations before convergence. Again Multinest and EMCEE are indicated in blue and red, respectively. The values shown on the top of the 1D PDF are median+/-[14,86]percentiles (maximum likelihood position)

I've also updated the EMCEE results with the last 5892 evaluations. Clearly we see similar posterior distribution with MultiNest being broader than EMCEE result. This is something worth looking into details.

image

**Comment by qyx268** [comment]
@BradGreig  I remember before when I showed you the comparison between Emcee and MultiNest using 5 parameters, it's also true that the posteriors look similar with MultiNest being a bit broader. You had the concern about sampling -- the number of points used to calculate the posterior is not large enough (I was using 7663 for the 5-parameter runs). I think, in this 3-parameter result where 1000 lives points were adopted and ~6000 points are used in the posterior, the factor that the comparison has the same behaviour indicates that Emcee might be over converged?

File: src/py21cmmc/likelihood.py
Comment (BradGreig) [inline src/py21cmmc/likelihood.py]: This should be 1420.4

I know it won't change any results, but it should at least be accurate. Probably should explicitly be 1.0 not 1, not sure how much is matters in Python for adding different variable types (or if it is intelligent enough to know that its a double)

File: src/py21cmmc/likelihood.py
Comment (BradGreig) [inline src/py21cmmc/likelihood.py]: You need to be careful when using this approximation for asymmetric errors. You need to put in checks that the denominator doesn't go negative

**Comment by BradGreig** [comment]
@qyx268 It may not be that its over converged, but perhaps dependent on what are the assumed initial conditions for EMCEE. I believe the default is to sample the initial conditions around the locations of the fiducial parameters. If you were to use completely random initial conditions instead, I wonder if the results would change slightly. 

In any case for nested sampling you must sample additional models in order to better fill out the posteriors for drawing contours. To me this comparison is pretty good.

File: src/py21cmmc/likelihood.py
Comment (qyx268) [reply src/py21cmmc/likelihood.py]: can I confirm with you that it is 1419.4697 or 1420.4?

Comment (BradGreig) [reply src/py21cmmc/likelihood.py]: It is 1420.405751...

Comment (qyx268) [reply src/py21cmmc/likelihood.py]: of course... now I'm wondeing why I've been using 1419 since the beginning ...

**Comment by steven-murray** [comment]
OK, let's wait until #19 is merged, and we can merge this one in v3.1 or so. Of course, you're more than welcome to use it for applications before then, with any hack you have for writing out blobs!

If we want to support different samplers into the future (emcee, multinest, polychord,...), it may be good to have a generic way of saving parameters/blobs from within the likelihood itself, in case the particular sampler doesn't expose their parameter sampling method. The thing is, the idea _should_ be to subclass the `CosmoHammerSampler` class for each kind of sampler. This can still be done (even for multinest), but because multinest is written in fortran(?), we don't have access to the actual parameters/outputs on each iteration -- it just runs away and does its thing. Polychord does the same thing.

**Comment by qyx268** [comment]
great, can someone review #19 please, is there anything else I need to do?

**Comment by steven-murray** [comment]
@qyx268 where's this PR at? If we update the branch with the base branch, could we merge it in?

**Comment by qyx268** [comment]
I have the impression that we've postponed it until outputting within likelihood is included. However, this PR does allow users to have access to parameters in the posterior. We can merge it now and users can hack the likelihoods for output if needed. 

I just merged the master into this branch. But it seems that, on Travis, `external_table_path`
is not correctly set, leading crashes in most tests

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: You can put them in a "Other Parameters" section -- remember that these docstrings are parsed by a program to convert to html documentation. The parser knows about the Parameters section and an "Other Parameters" section, and can format them appropriately.

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: I think I would favour the latter, actually. Let's just add an option for `run_mcmc` like `sampler`, with `emcee` as the default, but one could specify `"multinest"` if they want.

**Comment by steven-murray** [comment]
@qyx268 OK thanks for the update! I say we merge it without the ability to do the full outputting within the likelihood, and create an issue for that right now, so we don't forget to add the feature later. 

You should also make a (small) tutorial showing how to use 21CMMC with multinest, since it's a pretty big feature.

The `external_table_path` thing is something that I think has happened before. I'll try to look into it, but it might not be til next week unless I can find a spare moment.

File: src/py21cmmc/mcmc.py
Comment (qyx268) [reply src/py21cmmc/mcmc.py]: Then for input parameters that are not shared by emcee and multinest, I think I will need to put them into 
`mcmc_options={continue_sampling':True, 'reuse_burnin':True, 'log_level_21CMMC':None, 'sampler_cls':CosmoHammerSampler}` and 
`multinest_options={'log_level_21CMMC': None, 'n_live_points': 100, 'importance_nested_sampling': True, 'sampling_efficiency': 0.8, 'evidence_tolerance': 0.5, 'max_iter': 50, 'multimodal': True, 'write_output': True, 'resume': True}`

However, this will change some existing scripts for using the emcee sampler. Do you have other suggestions?

**Comment by qyx268** [comment]
> @qyx268 OK thanks for the update! I say we merge it without the ability to do the full outputting within the likelihood, and create an issue for that right now, so we don't forget to add the feature later.

okay!

> You should also make a (small) tutorial showing how to use 21CMMC with multinest, since it's a pretty big feature.

yes, I will include a tutorial for this feature

> The `external_table_path` thing is something that I think has happened before. I'll try to look into it, but it might not be til next week unless I can find a spare moment.

thx!

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: No, I think all of the new multinest options can be part of the existing `**mcmc_options`, and you can give them defaults in the function body by doing something like

```python
n_live_points = mcmc_options.get('n_live_points', 100)
```

Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: Also, 'resume' does the same thing as 'continue_sampling', so that's one less parameter.

Comment (qyx268) [reply src/py21cmmc/mcmc.py]: done. but we might need to consider renaming `run_mcmc` because multinest is not strictly speaking an mcmc.

**Comment by qyx268** [comment]
I'm not sure if we want to add a test for this feature because we need to instead MultiNest and PyMultiNest to use this feature...

**Comment by steven-murray** [comment]
I think we definitely want to test it -- just at a very basic level. Those packages should hopefully be installable fairly easily, right? We just have to write down the install steps in the travis.yaml.

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [reply src/py21cmmc/mcmc.py]: I think it's close enough that no-one is going to care :-)

**Comment by qyx268** [comment]
@steven-murray okay, they are included now and the test passes. Could you also help check the test and see if it makes sense? 

It's easy to include `pymultinest` in `setup.py`. Do you have any suggestions on building `MultiNest` when installing `py21cmmc`? You can take a look at the travis file for its installation.

File: src/py21cmmc/mcmc.py
Comment (steven-murray) [inline src/py21cmmc/mcmc.py]: Let's point the user to https://johannesbuchner.github.io/PyMultiNest/install.html#id4 here so they know how to install it.

**Review by steven-murray — APPROVED** [review]
OK, I think all of this is good. There seems to be an extra jupyter notebook in `src/py21cmmc/MultiNest.ipynb` that shouldn't be there, so please remove that.

The test errors that remain must be due to the `external_table_path`?

File: src/py21cmmc/mcmc.py
Comment (qyx268) [reply src/py21cmmc/mcmc.py]: ok

**Comment by qyx268** [comment]
> The test errors that remain must be due to the `external_table_path`?

this is the error
image

I think it's either the `external_table_path` has been wrongly set to `/home/travis/.21cmfast/` (see `EXTERNALTABLES` in `src/py21cmfast/inputs.py`) or `External_tables` was not correctly copied into `$HOME/.21cmfast` during installation.

**Comment by qyx268** [comment]
hmm... I'm not familiar with tox... After a few explorations, I manage to put the installation of MultiNest in tox.ini. But it seems that we cannot set LD_LIBRARY_PATH in tox.ini, which is required by MultiNest. Therefore, I have to put it in .travis.yml.

But in fact, MultiNest is not a python package. So we can actually put all of the steps in .travis.yml (I had one build here: https://travis-ci.com/github/21cmfast/21CMMC/jobs/352959474)

I'm not sure which one is the optimal way. Actually, do you know any other better or more commonly used setup? @steven-murray

**Comment by steven-murray** [comment]
@qyx268 I'll try giving this a go tomorrow. Travis/tox can be a bit of a pain for complex stuff.

**Comment by steven-murray** [comment]
@qyx I think the way you've done it is fine, and it looks like it works OK. This stuff is tricky, and I don't know of a necessarily "better" way to do it. I'm thinking perhaps we might just do away with tox, because I for one pretty much never test different python versions on my local system -- I let travis do that! I'm happy for you to merge this now if you're happy with it.

Thanks for the awesome work!

**Pull request merged**
